crates/autopilot/src/infra/solvers/authenticator.rs

```diff
@@ -0,0 +1,91 @@
+use {
+    crate::{
+        domain::{self, eth},
+        infra::blockchain::{
+            contracts::{deployment_address, Contracts},
+            ChainId,
+        },
+    },
+    ethcontract::{dyns::DynWeb3, GasPrice},
+};
+
+#[allow(dead_code)]
+#[derive(Debug, Clone)]
+pub struct Authenticator {
+    /// The authenticator contract that decides which solver is allowed to
+    /// submit settlements.
+    authenticator: contracts::GPv2AllowListAuthentication,
+    /// The safe module that is used to provide special role to EOA.
+    authenticator_role: contracts::Roles,
+    /// The EOA that is allowed to remove solvers.
+    authenticator_eoa: ethcontract::Account,
+}
+
+///  Authenticator specific addresses
+pub struct Addresses {
+    pub authenticator_eoa: eth::H256,
+}
+
+impl Authenticator {
+    /// Creates an authenticator which can remove solvers from the allow-list
+    pub async fn new(
+        web3: DynWeb3,
+        chain: ChainId,
+        contracts: Contracts,
+        addresses: Addresses,
+    ) -> Self {
+        let authenticator_role = contracts::Roles::at(
+            &web3,
+            deployment_address(contracts::Roles::raw_contract(), &chain).expect("roles address"),
+        );
+
+        Self {
+            authenticator: contracts.authenticator().clone(),
+            authenticator_role,
+            authenticator_eoa: ethcontract::Account::Offline(
+                ethcontract::PrivateKey::from_raw(addresses.authenticator_eoa.0).unwrap(),
```

Comment:
Since the `eoa` must be passed as a private key `authenticator_pk` seems like a more appropriate name.